CIP-0XXX/README.md

```diff
@@ -0,0 +1,561 @@
+  StMessageIdsBlocking:::Green --> StDone:::Black : MsgDone
+  StIdle:::Blue --> StMessages:::Green : MsgRequestMessages
+  StMessages:::Green --> StIdle:::Blue : MsgReplyMessages
+   
+```
+
+#### Protocol messages
+
+* **MsgInit**: Initial message of the protocol.
+* **MsgRequestMessageIdsNonBlocking(ack,req)**: The server asks for new message ids and acknowledges old ids. The client immediately replies (possible with an empty list).
+* **MsgRequestMessageIdsBlocking(ack,req)**: The server asks for new messages ids and acknowledges old ids. The client will block until new messages are available.
+* **MsgReplyMessageIds([(id,size)])**: The client replies with a list of available messages. The list contains pairs of message ids and the corresponding size of the message in bytes. In the blocking case the reply is guaranteed to contain at least one message. In the non-blocking case, the reply may contain an empty list.
+* **MsgRequestMessages([ids])**: The server requests messages by sending a list of message-ids.
+* **MsgReplyMessages([messages])**: The client replies with a list messages.
+* **MsgClientDone**: The client terminates the mini-protocol.
+
+#### Transition table
+
+| From state | Message | Parameters | To State|
+|--------|------|----|----| 
+| StInit | MsgInit |  | StIdle |
+| StIdle | MsgRequestMessageIdsBlocking | ack,req | StMessageIdsBlocking |
+| StMessageIdsBlocking | MsgReplyMessageIds | [(id,size)] | StIdle |
+| StIdle | MsgRequestMessageIdsNonBlocking | ack,req | StMessageIdsNonBlocking |
+| StMessageIdsNonBlocking | MsgReplyMessageIds | [(id,size)] | StIdle |
+| StIdle | MsgRequestMessages | [ids] | StMessages |
+| StMessages | MsgReplyMessages | [messages] | StIdle |
+| StMessageIdsBlocking | MsgDone |  | StDone |
+
+#### CDDL encoding specification
+
+
+```cddl
+ 1
+ 2  messageSubmissionMessage
+ 3    = msgInit
+ 4    / msgRequestMessageIds
+ 5    / msgReplyMessageIds
+ 6    / msgRequestMessages
+ 7    / msgReplyMessages
+ 8    / msgDone
+ 9 
+10  msgInit = [0]
+11  msgRequestMessageIds = [1, isBlocking, messageCount, messageCount]
+12  msgReplyMessageIds = [2, [ *messageIdAndSize ] ]
+13  msgRequestMessages = [3, messageIdList ]
+14  msgReplyMessages = [4, ]
+15  msgDone = [5, ]
+16
+17  isBlocking = false / true
+18  messageCount = word16
+19  messageId = bstr
+20  messageBody = bstr
+21  messageIdAndSize = [ messageId, messageSizeInBytes ]
+22  messageIdList = [ * messageId ]
+23  messageList = [ * message ]
+24  messageSizeInBytes = word32
+25  kesSignature = bstr
+26  operationalCertificate = bstr
+27  blockNumber = word32
+28  
+29  message = [
+30    messageId,
+31    messageBody,
+32    blockNumber,
+33    kesSignature,
+34    operationalCertificate
+35  ]
+30
+```
+
+### Client and server implementation
+
+This mini-protocol is designed with two goals in mind:
+* diffuse messages with high efficiency
+* protect from asymmetric resource atatckes from the message consumer against the message provider
```

Comment:
It's more related to the fact that when a peer connects to another peer it serves requests to that peer instead of initiating requests to it (which is why the server has agency in the state diagram).